### ISSUE TITLE
Monsters cast beams and waves through wall

### ISSUE THREAD

**Issue by LuisPro** [otland/forgottenserver#1690]
When monster is in front to wall (anything, any BLOCKSOLID wall, included magic wall) and he is blocked by player..
Monster will cast beam and wave through wall!
http://i.imgur.com/63nsVoh.png

**Comment by marksamman** [comment]
This might be working as intended, because the location where the player is standing is allowing it to go through the wall (and creatures don't block waves). I'm not entirely sure though, so I'll leave this open for discussion.

**Comment by TheSumm** [comment]
I did not check for any more fitting location, but this should be the same circumstances tested on real tibia.

http://i.imgur.com/HLPOFzh.jpg

**Comment by LuisPro** [comment]
@TheSumm Yes, it should look like on your screen. Not like my 1st screen.

Now we have this: http://i.imgur.com/Al3TYJo.png
It is 100% wrong.

**Comment by degolaot** [comment]
Same in warzone entrance:

http://prntscr.com/9shzvo

**Comment by BenDol** [comment]
Anyone know why its doing this for monsters but not for players?

**Comment by LuisPro** [comment]
Becouse for players its blocked in Source. For monsters is not.

**Comment by TheSumm** [comment]
For players it checks if the tile infront of the player blocks the spells, **which should be removed** as you may cast all spells even when facing walls now: http://i.imgur.com/CJjLATQ.gifv
https://github.com/otland/forgottenserver/blob/master/src/spells.cpp#L696-L701

I don't know why the area check fails for monsters despite being behind a wall.

**Comment by BenDol** [comment]
Could be an issue with isSightClear?

**Comment by whitevo** [comment]
Well not really a fix to problem just a work around I use, maybe you want to go for it too.

Monsters cast spells in onThink() creaturescript and you make custom path finder. Which will take shortest path to target and check each position, does it block spell.

This way you can easily add custom things like: if fireball is shot trough fire field it gets stronger or if it goes trough magic wall it changes direction or if its mirror then its reflected, etc.

**Comment by LuisPro** [comment]
still unconfirmed? okz

**Comment by wgrriffel** [comment]
At this moment on tibia you can cast spells when facing walls but it don't realize the spell area that are facing the wall, i just wonder if this is a feature that @marksamman would like to have here.

**Comment by Marofdp** [comment]
its work normal free sqm see dragon using more magic wall dragon not use fire wave only fire bomb on free sqm :) on me its not bug ! better ways go to find other rlly critical bugs ..

i test this on rl / hexera  and my server its same effect !

**Comment by wgrriffel** [comment]
@Marofdp no, on real tibia all sqms instead that are behind the wall would get the damage area, here its going to all sqms, it's a bug.

**Comment by LuisPro** [comment]
@Marofdp english pls

**Comment by kamil-karkus** [comment]
Is player able to cast area spells over trees, bushes, etc...?

**Comment by kamil-karkus** [comment]
http://imgur.com/mz0MfYB
could anybody tell if it's expected behaviour?

**Comment by BenDol** [comment]
@kamil-karkus Yeah that looks to be the expected behavior

**Comment by kamil-karkus** [comment]
@BenDol thanks

http://imgur.com/lQcZtqm (dragon firebomb)
http://imgur.com/w0DuhlP (player firebomb)
http://imgur.com/ez9CgvA (player waves)
http://imgur.com/G5qZuO7 (dragon waves)

Please confirm if it works fine so I can create pull request tomorrow

**Comment by kamil-karkus** [comment]
How should behave distance weapon? it should reach places reached by 'exevo tera hur' ? 
or not? I can't test it on real tibia so need your help.

**Comment by TheSumm** [comment]
@kamil-karkus Neither the tree nor the small stone should be blocking the dragon's wave nor the player's wave.

**Comment by wgrriffel** [comment]
@kamil-karkus i've recorded 10 mins my screen waving through walls, take a look
https://youtu.be/UrXbFjzNTGg

**Comment by Zbizu** [comment]
Might be a duplicate of #2891
